File: runtimes/compose/src/main/java/com/styleconverter/runtime/columns/GapDecorationSegments.kt
package com.styleconverter.runtime.columns

// css-gap-decorations-1 §3-§5 — the flex gap-decoration SEGMENT MODEL
// (wave 24, lane GAPS-A). Pure geometry: rectangles in, rectangles out.
//
// Every number in the comments below was re-derived from the WPT
// REFERENCE documents for css-gaps/flex (fixtures/wpt/css-gaps/
// flex__flex-gap-decorations-NNN__ref.json — the refs are the browser's
// own answer expressed as absolutely-positioned boxes, so they pin the
// geometry exactly). The shared 003-family layout used throughout is:
// 2px border, 170px content box at origin (2,2), column-gap 10, row-gap
// 10, wrap, item widths 50/50/50/100/50/50/50/50 × 50px tall, giving
//   line 1 items x [2,52] [62,112] [122,172], cross [2,52]
//   line 2 items x [2,102] [112,162],         cross [62,112]
//   line 3 items x [2,52] [62,112] [122,172], cross [122,172]
//
// THE MODEL, and what pins each rule:
//  (b) a within-line (column) rule is one segment per adjacent item pair,
//      of the declared width, CENTERED in the item-to-item gap, spanning
//      that LINE's cross extent — pinned by 007 (line extent, not item
//      extent), 008 (five segments, painted outside the container box),
//      012 (`column-over-row` makes the extent observable: the 2px red
//      rules are EXACTLY [2,52] and [122,172], so the segment does NOT
//      run into the row gaps).
//  (c) a between-line (row) rule spans the container CONTENT BOX along
//      the main axis, centered in the inter-line gap — 003/010/011/012
//      all draw it as a full 170px bar at content-box left.
//  (d) `intersection` cuts a rule at every crossing GAP (not at the
//      crossing rule's width) — 034 is decisive: 5px rules, 90px gaps,
//      and the ref cuts the row rule by the full 90px columns.
//  (e) `*-rule-inset` shortens both ends, negative lengthens — 013 (+2px)
//      and 011 (-2px, which escapes the content box).
//  (f) `rule-overlap` is paint ORDER only — 012.

/**
 * Builds the ordered list of rule rectangles for one flex container.
 *
 * Inputs are all in the container's CONTENT-BOX coordinate space, real
 * pixels. The caller (the draw hook) is responsible for translating the
 * placed item bounds into that space; this object does no unit work.
 */
object GapDecorationSegments {

    /**
     * @param config        the container's resolved gap-decoration config.
     * @param items         placed flex-item rectangles, in placement order.
     * @param contentBox    the container's content box, origin-relative
     *                      (normally (0,0,w,h)); supplies the span of the
     *                      between-line rules.
     * @param mainHorizontal true for `flex-direction: row|row-reverse` in
     *                      a horizontal writing mode. Determines which
     *                      physical family each gap belongs to: with a
     *                      horizontal main axis the item-to-item gaps are
     *                      horizontal → COLUMN rules, and the inter-line
     *                      gaps are vertical → ROW rules; with a vertical
     *                      main axis the two swap.
     * @return segments in PAINT ORDER (first painted first).
     */
    fun build(
        config: GapDecorationConfig,
        items: List<GapRect>,
        contentBox: GapRect,
        mainHorizontal: Boolean
    ): List<GapSegment> {
        // Inert configs short-circuit: this is the gate that keeps every
        // existing fixture byte-identical (nothing declares these props).
        if (!config.active || items.isEmpty()) return emptyList()

        val lines = GapDecorationLines.toLines(
            GapDecorationLines.groupIntoLines(items, mainHorizontal), mainHorizontal
        )
        // Which physical family owns which gap kind (see @param above).
        val withinAxis = if (mainHorizontal) GapAxis.COLUMN else GapAxis.ROW
        val betweenAxis = if (mainHorizontal) GapAxis.ROW else GapAxis.COLUMN
        val interLine = GapDecorationLines.betweenLineGaps(lines)

        val within = withinLineSegments(
            config.specFor(withinAxis), withinAxis, lines, interLine, mainHorizontal
        )
        val between = betweenLineSegments(
            config.specFor(betweenAxis), betweenAxis, lines, contentBox, mainHorizontal
        )

        // (f) rule-overlap decides which family lands on top; painting is
        // last-wins, so the loser goes first. Initial value is
        // ROW_OVER_COLUMN, so columns paint first (WPT 012 flips it).
        val columns = if (withinAxis == GapAxis.COLUMN) within else between
        val rows = if (withinAxis == GapAxis.COLUMN) between else within
        return when (config.overlap) {
            GapRuleOverlap.ROW_OVER_COLUMN -> columns + rows
            GapRuleOverlap.COLUMN_OVER_ROW -> rows + columns
        }
    }

    /**
     * (b) One segment per adjacent item pair per line.
     *
     * [interLineGaps] are handed in as potential cuts so `intersection` is
     * implemented symmetrically for both families. In flex they can never
     * actually cut anything — a line's cross extent lies strictly between
     * its neighbouring inter-line gaps — so this is a PROVEN no-op here,
     * asserted by a unit test rather than assumed. Grid/multicol would use
     * the same operator with real cuts.
     */
    private fun withinLineSegments(
        spec: GapRuleSpec,
        axis: GapAxis,
        lines: List<GapFlexLine>,
        interLineGaps: List<GapInterval>,
        mainHorizontal: Boolean
    ): List<GapSegment> {
        if (!spec.paints) return emptyList()
        val cuts = if (spec.breakMode == GapRuleBreak.INTERSECTION) interLineGaps else emptyList()
        val out = mutableListOf<GapSegment>()
        for (line in lines) {
            for (gap in line.mainGaps) {
                // Declared width, centered in the item-to-item gap.
                val band = GapIntervals.band(gap, spec.effectiveWidthPx)
                // Length runs along the CROSS axis, over the line's extent.
                for (piece in GapIntervals.subtract(line.cross, cuts)) {
                    val len = GapIntervals.inset(piece, spec.insetPx) ?: continue
                    // Band is main-axis, length is cross-axis: assemble in
                    // physical (x, y) order according to the main axis.
                    val rect = if (mainHorizontal) GapIntervals.rect(band, len)
                    else GapIntervals.rect(len, band)
                    out.add(GapSegment(rect, axis))
                }
            }
        }
        return out
    }

    /**
     * (c)+(d) One rule per adjacent line pair, spanning the content box
     * along the main axis and cut — under `intersection` — by the union of
     * BOTH adjacent lines' item-to-item gaps.
     *
     * The two-line union is the part that is easy to get wrong and is
     * pinned three times: WPT 009 (cuts {[52,62],[112,122]} ∪ {[102,112]}
     * → segments [2,52] [62,102] [122,172]), WPT 034 (four segments on a
     * 600px box), WPT 050 (six segments with `justify-content:
     * space-between` opening unequal gaps on the two lines).
     */
    private fun betweenLineSegments(
        spec: GapRuleSpec,
        axis: GapAxis,
        lines: List<GapFlexLine>,
        contentBox: GapRect,
        mainHorizontal: Boolean
    ): List<GapSegment> {
        if (!spec.paints || lines.size < 2) return emptyList()
        // The span every between-line rule is drawn across.
        val span = GapDecorationLines.mainOf(contentBox, mainHorizontal)
        val out = mutableListOf<GapSegment>()
        for (i in 0 until lines.size - 1) {
            val gap = GapInterval(lines[i].cross.end, lines[i + 1].cross.start)
            if (gap.isEmpty) continue
            val band = GapIntervals.band(gap, spec.effectiveWidthPx)
            val cuts = if (spec.breakMode == GapRuleBreak.INTERSECTION) {
                lines[i].mainGaps + lines[i + 1].mainGaps
            } else {
                emptyList()
            }
            for (piece in GapIntervals.subtract(span, cuts)) {
                // The inset applies to EACH surviving piece's two ends —
                // WPT 009/034 set it to 0 so the pieces end exactly on the
                // cut edges, which is what that convention reproduces.
                val len = GapIntervals.inset(piece, spec.insetPx) ?: continue
                // Band is cross-axis here, length is main-axis.
                val rect = if (mainHorizontal) GapIntervals.rect(len, band)
                else GapIntervals.rect(band, len)
                out.add(GapSegment(rect, axis))
            }
        }
        return out
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/columns/GapDecorationSegments.swift
//
//  GapDecorationSegments.swift
//  StyleEngine/columns — wave 24, lane GAPS-I. PURE, view-free.
//
//  Item frames + a GapDecorationsConfig → the exact rects css-gaps-1
//  says to paint. Byte-parallel with the Compose twin
//  (…/columns/GapDecorationSegments.kt): same `build` entry point over
//  the shared GapDecorationLines / GapIntervals primitives.
//
//  THE RENDER MODEL, re-derived from FRESH renders of the WPT css-gaps
//  flex refs (tools/wpt/css/css-gaps/flex/flex-gap-decorations-0NN-ref
//  .html through the capture-browser-ref.mjs recipe — same puppeteer
//  BROWSER_LAUNCH_ARGS, same injected canvas CSS; the committed refs are
//  stale). Ref coordinates below are the flexbox BORDER-box space the
//  refs' absolutely-positioned decoration divs live in; this file works
//  in CONTENT-box space, i.e. those numbers minus the 2px border.
//
//   (a) LINE GROUPING → GapDecorationLines (see that file).
//   (b) ITEM-GAP RULES (the column rules of a row-direction container)
//       sit centred in each adjacent-item gap band and span the LINE's
//       cross extent — ref 003's third column div is x102–112 × y62–112
//       = line 2 exactly.
//   (c) LINE-GAP RULES (row rules) sit centred in each adjacent-line gap
//       band and span the container's whole content main extent — ref
//       003's row divs are [2,52,170,10].
//   (d) BREAKS. SPANNING_ITEM (the CSS initial) lets an item-gap rule
//       CROSS the line gap when the neighbouring line has a band at the
//       same position — refs 001/002 paint one column rule down the
//       container's full height. A spanning item severs it AT THE LINE
//       EDGE: ref 012 turns `rule-overlap: column-over-row` on, which
//       makes those ends visible, and its column divs are y2–52 /
//       y122–172 with no bleed into the row gap. INTERSECTION cuts a
//       line-gap rule over BOTH neighbouring lines' item-gap bands —
//       ref 009's row children are x2–52 / 62–102 / 122–172.
//   (e) INSETS shrink (positive) or EXTEND (negative) both ends — ref
//       011's `column-rule-inset: -2px` grows 50px runs to 54px, the
//       first starting 2px ABOVE the content box.
//   (f) RULE-OVERLAP orders the two families (ref 003 row-over-column,
//       ref 012 column-over-row).
//
//  Test 006 is out of scope: `writing-mode: vertical-lr` is the wall.
//

import SwiftUI

/// One paintable rule run. `isRowRule` picks the family for the
/// `rule-overlap` ordering AND the line orientation (a column rule is
/// always a vertical line, a row rule always horizontal). The rect is in
/// the container's CONTENT-box space.
struct GapRuleSegment: Equatable {
    /// Where to paint, content-box relative.
    var rect: CGRect
    /// Resolved colour; nil = inherit (`currentColor`).
    var color: Color?
    /// Line style — strokes go through the border machinery.
    var style: BorderStyleValue
    /// True for `row-rule-*`, false for `column-rule-*`.
    var isRowRule: Bool
}

enum GapDecorationSegments {

    // MARK: - (b)+(d)+(e) item-gap rules

    /// Rules painted in the gaps BETWEEN ITEMS of a line. For a
    /// row-direction container these are the `column-rule-*` runs.
    /// Returns (main, cross) interval pairs in content-box coordinates.
    static func itemGapRuns(lines: [GapLine],
                            spec: GapRuleSpec,
                            contentCross: GapSpan) -> [(main: GapSpan, cross: GapSpan)] {
        // No ink → no runs (keeps the family free when a container
        // declares, say, only a colour).
        // `effectiveWidthPx`, not `widthPx`: an undeclared width is the
        // CSS initial `medium` (3px), not "no rule".
        guard spec.paints else { return [] }
        let w = spec.effectiveWidthPx
        // ONE RUN PER LINE, ALWAYS — an item-gap rule NEVER crosses a
        // line gap, whatever `*-rule-break` says.
        //
        // This was measured in Chrome 150, not inferred from the refs.
        // Take flex-gap-decorations-002's 2×2 layout, narrow BOTH rules
        // to 2px so neither can mask the other, and scan the column band
        // (x54) down through the row gap (y50…60):
        //     y49 red · y50–53 WHITE · y54–55 blue · y56–59 WHITE · y60 red
        // The column rule stops dead at each line edge; only the row
        // rule's own 2px band has ink in between. Repeating the scan
        // with `rule-overlap: column-over-row` and with
        // `column-rule-break: intersection` gives byte-identical
        // results. (Probe: tools/visual/_skeptic/probe-scan.mjs.)
        //
        // The refs cannot see this. agnostic/gap-decorations-001-ref
        // (the match for flex 002) draws its column-gap div `height:
        // 110px`, i.e. full-container, and flex/…-015-ref draws its row
        // rules as two 50px runs with a hole — but in BOTH the crossing
        // square is repainted by the other family, so a merged and an
        // unmerged rule are pixel-identical there. Reading the ref's
        // markup as the model is exactly the trap the 55px bleed in
        // ref 003 already demonstrates. The Compose twin
        // (GapDecorationSegments.withinLineSegments) is per-line too.
        //
        // `contentCross` is therefore unused; it stays in the signature
        // for the still-unmapped CSS `none` keyword (44 occurrences under
        // tools/wpt/css/css-gaps/, currently rejected by the converter
        // into GenericProperty and never reaching this file), which is
        // the one value that could want a container-wide extent.
        _ = contentCross
        // `*-rule-break` has nothing left to decide for THIS family in
        // flex: `normal`/`spanning-item` break at spanning items, which
        // css-flexbox-1 §5.2 forbids, and `intersection`'s crossings are
        // the line gaps a per-line run already stops at. Named, not
        // silently dropped — grid will need the distinction.
        _ = spec.breakMode
        return lines.flatMap { line -> [(main: GapSpan, cross: GapSpan)] in
            let extent = GapSpan(start: line.crossStart, end: line.crossEnd)
            // The inset moves both ends; negative EXTENDS them past the
            // line edge (ref 011's `column-rule-inset: -2px`).
            guard let cross = GapIntervals.inset(extent, insetPx: spec.insetPx) else {
                return []
            }
            return line.gapBands.map {
                (GapIntervals.band(gap: $0, widthPx: w), cross)
            }
        }
    }

    // MARK: - (c)+(d)+(e) line-gap rules

    /// Rules painted in the gaps BETWEEN LINES. For a row-direction
    /// container these are the `row-rule-*` runs.
    static func lineGapRuns(lines: [GapLine],
                            spec: GapRuleSpec,
                            contentMain: GapSpan) -> [(main: GapSpan, cross: GapSpan)] {
        // Same no-ink short circuit; one line has no line gaps.
        guard spec.paints, lines.count > 1 else { return [] }
        // Same `medium` initial as the item-gap family.
        let w = spec.effectiveWidthPx
        var out: [(main: GapSpan, cross: GapSpan)] = []
        for i in 0..<(lines.count - 1) {
            // The band between two adjacent line boxes.
            let gap = GapSpan(start: lines[i].crossEnd, end: lines[i + 1].crossStart)
            // A zero/negative band (touching lines) decorates nothing.
            guard gap.isPositive else { continue }
            let cross = GapIntervals.band(gap: gap, widthPx: w)
            // Default run: the whole content main extent — refs 003/010/
            // 011 all pin the row rule at the full 170px content width.
            var runs = [contentMain]
            if spec.breakMode == .intersection {
                // Ref 009: cut over the union of BOTH neighbouring lines'
                // item-gap bands — the crossing GAPS, not the crossing
                // rule widths (a rule may be narrower than its gap).
                runs = GapIntervals.subtract(
                    base: contentMain,
                    cuts: lines[i].gapBands + lines[i + 1].gapBands)
            }
            // SPANNING_ITEM would additionally break where an item spans
            // the LINE gap; a flex item can never span two flex lines
            // (css-flexbox-1 §5.2), so there is nothing to cut. Stated
            // rather than silently skipped.
            for r in runs {
                if let main = GapIntervals.inset(r, insetPx: spec.insetPx) {
                    out.append((main, cross))
                }
            }
        }
        return out
    }

    // MARK: - (f) assembly + paint order

    /// The full segment list for one container, BACK TO FRONT.
    /// `contentSize` is the container's CONTENT box; `items` are the
    /// measured item frames in that same space.
    static func build(items: [CGRect],
                      contentSize: CGSize,
                      mainHorizontal: Bool,
                      config: GapDecorationsConfig) -> [GapRuleSegment] {
        // Cheapest possible bail — the state every component in the
        // committed corpus is in (no *-rule-* resolves to ink).
        guard config.isActive else { return [] }
        let lines = GapDecorationLines.lines(from: items, mainHorizontal: mainHorizontal)
        guard !lines.isEmpty else { return [] }
        // Content extents split into (main, cross) for this axis.
        let mainExtent = GapSpan(start: 0,
                                 end: mainHorizontal ? contentSize.width : contentSize.height)
        let crossExtent = GapSpan(start: 0,
                                  end: mainHorizontal ? contentSize.height : contentSize.width)
        // Which family owns which gap kind: item gaps run ALONG the main
        // axis, so in a row-direction container they are the COLUMN gaps
        // and in a column-direction container the ROW gaps.
        let itemSpec = mainHorizontal ? config.column : config.row
        let lineSpec = mainHorizontal ? config.row : config.column
        // Item-gap family.
        let itemSegs = itemGapRuns(lines: lines, spec: itemSpec, contentCross: crossExtent)
            .map { GapRuleSegment(rect: place($0, mainHorizontal: mainHorizontal),
                                  color: itemSpec.color,
                                  style: itemSpec.style ?? .solid,
                                  isRowRule: !mainHorizontal) }
        // Line-gap family.
        let lineSegs = lineGapRuns(lines: lines, spec: lineSpec, contentMain: mainExtent)
            .map { GapRuleSegment(rect: place($0, mainHorizontal: mainHorizontal),
                                  color: lineSpec.color,
                                  style: lineSpec.style ?? .solid,
                                  isRowRule: mainHorizontal) }
        // rule-overlap: the winning family goes LAST so it paints on top.
        let (column, row) = mainHorizontal ? (itemSegs, lineSegs) : (lineSegs, itemSegs)
        return config.overlap == .rowOverColumn ? column + row : row + column
    }

    // MARK: - helpers

    /// (main, cross) → a rect in the container's coordinate space.
    private static func place(_ run: (main: GapSpan, cross: GapSpan),
                              mainHorizontal: Bool) -> CGRect {
        // Horizontal main → main is x, cross is y; vertical main swaps.
        mainHorizontal
            ? GapIntervals.rect(horizontal: run.main, vertical: run.cross)
            : GapIntervals.rect(horizontal: run.cross, vertical: run.main)
    }
}
